client/components/data/query-user-purchases/index.jsx

```diff
@@ -29,4 +29,16 @@ function QueryUserPurchases() {
 	return null;
 }
 
+export const useQueryUserPurchases = () => {
+	const userId = useSelector( ( state ) => getCurrentUserId( state ) );
+	const isRequesting = useSelector( ( state ) => state.purchases.isFetchingUserPurchases );
+	const hasLoaded = useSelector( ( state ) => state.purchases.hasLoadedUserPurchasesFromServer );
+	const reduxDispatch = useDispatch();
+
+	if ( ! userId || isRequesting || hasLoaded ) {
+		return;
+	}
+	reduxDispatch( fetchUserPurchases( userId ) );
```

Comment:
I had a difficult time with this one. Initially, I wanted to do it just like `useQuerySitePurchases`, and I tried to do like so, but:

This function is triggered many times, as it's inside the `use-site-copy` hook. (it does for each row item).

I was expecting that having a `useEffect` will actually prevent the dispatch of `fetchUserPurchaces` again, after `PURCHASES_USER_FETCH` action.

However this is not the case, and if I put this code inside a `useEffect`, the request is triggered for each row item, no matter what.

Moving this code:

```
if ( isRequesting || hasLoaded ) {
		return;
	}
```
inside `fetchUserPurchases`, solves the issue, but, I wasn't sure if we want to put that logic there.

Having said that, and because of the things I mentioned, I'm not sure that `useQuerySitePurchases` works as expected tbh.

Reply:
If the checks for whether the purchases data is inside the effect too, then wouldn't this prevent the repeated queries?

```ts
useEffect( () => {
	if ( ! userId || isRequesting || hasLoaded ) {
		return;
	}
	reduxDispatch( fetchUserPurchases( userId ) );
}, [ userId, isRequesting, hasLoaded, reduxDispatch ] );
```

The reason we want to put the fetch in an effect hook is that it moves it out of the high-priority part of the React lifecycle and into the low-priority part.

<img width="398" alt="CleanShot 2023-02-07 at 15 28 26@2x" src="https://user-images.githubusercontent.com/1500769/217132761-026c6d7c-b6e8-4795-8732-48037dfb779c.png">

[[source]](https://github.com/donavon/hook-flow)

Currently `fetchUserPurchases` is being called in the green part, but we want it at the end in the yellow part.

Reply:
First of all, thanks for the info!!! REALLY handy!
I was about to respond that I had tried it but it did not work. However, I thought, let's try that again to confirm that, as I had already tried it!
But miraculously it worked!! So I pushed that change! Thanks!